worldedit-core/src/main/java/com/sk89q/worldedit/function/mask/OffsetsMask.java

```diff
@@ -0,0 +1,172 @@
+            .collect(Collectors.toList())
+    );
+
+    private final Mask mask;
+    private final boolean excludeSelf;
+    private final int minMatches;
+    private final int maxMatches;
+    private final List<BlockVector3> offsets;
+
+    /**
+     * Create an OffsetsMask for a single offset.
+     *
+     * @param mask The mask to use
+     * @param offset The offset
+     * @return The offsets mask
+     */
+    public static OffsetsMask single(Mask mask, BlockVector3 offset) {
+        return new OffsetsMask(mask, false, 1, 1, ImmutableList.of(offset));
+    }
+
+    /**
+     * Create a new instance.
+     *
+     * @param mask the mask to test against
+     */
+    public OffsetsMask(Mask mask) {
+        this(mask, false);
+    }
+
+    /**
+     * Create a new instance.
+     *
+     * @param mask the mask to test against
+     * @param excludeSelf excludes blocks where the mask matches itself
+     */
+    public OffsetsMask(Mask mask, boolean excludeSelf) {
+        this(mask, excludeSelf, 1, Integer.MAX_VALUE);
+    }
+
+    /**
+     * Create a new instance.
+     *
+     * @param mask the mask to test against
+     * @param excludeSelf excludes blocks where the mask matches itself
+     * @param minMatches the minimum number of matches
+     * @param maxMatches the maximum number of matches
+     */
+    public OffsetsMask(Mask mask, boolean excludeSelf, int minMatches, int maxMatches) {
+        this(mask, excludeSelf, minMatches, maxMatches, OFFSET_LIST);
+    }
+
+    /**
+     * Create a new instance.
+     *
+     * @param mask the mask to test against
+     * @param excludeSelf excludes blocks where the mask matches itself
+     * @param minMatches the minimum number of matches
+     * @param maxMatches the maximum number of matches
+     * @param offsets the block offsets to test with
+     */
+    public OffsetsMask(Mask mask, boolean excludeSelf, int minMatches, int maxMatches, List<BlockVector3> offsets) {
+        checkNotNull(mask);
+        checkNotNull(offsets);
+        // Validate match args. No need to test maxMatches as it must be >=0 based on the conditions here.
+        checkArgument(minMatches <= maxMatches, "minMatches must be less than or equal to maxMatches");
+        checkArgument(minMatches >= 0, "minMatches must be greater than or equal to 0");
+        checkArgument(minMatches <= offsets.size(), "minMatches must be less than or equal to the number of offsets");
+        checkArgument(offsets.size() > 0, "offsets must have at least one element");
+
+        this.mask = mask;
+        this.excludeSelf = excludeSelf;
+        this.minMatches = minMatches;
+        this.maxMatches = maxMatches;
+        this.offsets = offsets;
+    }
+
+    /**
+     * Get the mask.
+     *
+     * @return the mask
+     */
+    public Mask getMask() {
+        return mask;
+    }
+
+    /**
+     * Get the offsets.
+     *
+     * @return the offsets
+     */
+    public List<BlockVector3> getOffsets() {
```

Comment:
```suggestion
    public ImmutableList<BlockVector3> getOffsets() {
```
It's recommended to do this if the API declares the list immutable. I would also recommend doing the copy on creation and storing it as the `offsets` field.